### ISSUE TITLE
Account's `auth_tx` function support

### ISSUE THREAD

**Issue by greenhat** [0xPolygonMiden/compiler#123]
Account should define a function `auth_tx` that is used to authenticate a transaction in the transaction script. See [tx script example](https://github.com/0xPolygonMiden/miden-base/blob/68c75e7a12037ab0e617a7eae578d8962d515fcd/mock/src/constants.rs#L44-L51) and it's usage in [p2id script test](https://github.com/0xPolygonMiden/miden-base/blob/68c75e7a12037ab0e617a7eae578d8962d515fcd/miden-tx/tests/integration/scripts/p2id.rs#L26-L127) 

Here is how it is defined in MASM:
```
use.miden::account
use.miden::tx
use.std::crypto::dsa::rpo_falcon512

# Slot in account storage at which the public key is stored.
const.PUBLIC_KEY_SLOT=0

#! Authenticate a transaction using the Falcon signature scheme
#! Stack: []
#! Output: []
#!
export.auth_tx_rpo_falcon512
    # Get commitments to created notes
    exec.tx::get_output_notes_hash
    # => [OUTPUT_NOTES_HASH, ...]

    # Get commitments to consumed notes
    exec.tx::get_input_notes_hash
    # => [INPUT_NOTES_HASH, OUTPUT_NOTES_HASH, ...]

    # Get current nonce of the account and pad
    exec.account::get_nonce push.0.0.0
    # => [0, 0, 0, nonce, INPUT_NOTES_HASH, OUTPUT_NOTES_HASH, ...]

    # Get current AccountID and pad
    exec.account::get_id push.0.0.0
    # => [0, 0, 0, account_id, 0, 0, 0, nonce, INPUT_NOTES_HASH, OUTPUT_NOTES_HASH, ...]

    # Compute the message to be signed
    # M = h(OUTPUT_NOTES_HASH, h(INPUT_NOTES_HASH, h(0, 0, 0, account_id, 0, 0, 0, nonce)))
    hmerge hmerge hmerge

    # Get public key from account storage at pos 0 and verify signature
    push.PUBLIC_KEY_SLOT exec.account::get_item
    # => [PUB_KEY, M]

    # Verify the signature against the public key and the message. The procedure gets as
    # inputs the hash of the public key and the hash of the message via the operand
    # stack. The signature is provided via the advice stack.The signature is valid if and
    # only if the procedure returns.
    exec.rpo_falcon512::verify
    # => []

    # Update the nonce
    push.1 exec.account::incr_nonce
    # => []
end
```
at https://github.com/0xPolygonMiden/miden-base/blob/68c75e7a12037ab0e617a7eae578d8962d515fcd/miden-lib/asm/miden/contracts/auth/basic.masm#L8-L51

To compile this function, we need to implement the following first:
- Expose `hmerge` Miden VM op (RPO hash function) in Rust via compiler intrinsics;
- Support for Miden's stdlib as a dependency (for `rpo_falcon512::verify`);

**Comment by greenhat** [comment]
> > Expose `hmerge` Miden VM op (RPO hash function) in Rust via compiler intrinsics;
> 
> I wonder how we'll expose this to the user in Rust. The ideal way would be via `miden` library so that it looks something like this:
> 
> ```rust
> use miden::crypto::hash::{Rpo256, RpoDigest};
> 
> // RpoDigest is basically just RpoDigest(Word)
> let a = RpoDigest::default();
> let b = RpoDigest::default();
> let c = Rpo256::merge(a, b); // this would execute hmerge instruction
> ```

Yes, that is pretty much how I imagined it. Maybe, with an additional high-level `fn Rpo256::from_bytes(bytes: Vec<u8>) -> Rpo256`, which will be implemented in terms of `Rpo256::merge`.

> 
> But I'm not sure if there is something blocking this.
> 
> Edit: If this simplifies things, maybe we should have a "dummy" procedure somewhere in `stdlib` which just calls `hmerge`?

I don't think it's needed.

**Comment by bitwalker** [comment]
> I wonder how we'll expose this to the user in Rust. The ideal way would be via miden library so that it looks something like this:

Compiler intrinsics (which will technically include some of the Miden Assembly instructions, such as `hmerge`) will be uniformly represented in Rust as `#[link_name = "miden.<intrinsic name recognized by the compiler>"] extern fn rust_facing_intrinsic_name(...) -> ..`. The compiler will replace any calls to such an intrinsic with an appropriate implementation during codegen.

The specific convention used for this is not specified yet, but that basic approach is going to be used for all intrinsics unless otherwise noted somewhere.
